### PR TITLE
[expo-updates][iOS] Allow native debugging of updates

### PR DESCRIPTION
**Pull request by douglowder**
# Why

Currently, if you have a project with expo-updates installed, you need to make a release build for expo-updates to be enabled (debug builds behave like normal RN project debug builds). This makes it somewhat difficult to test and debug updates when things aren’t working as expected.

To make this easier, we want to add a build setting on both platforms to enable updates in debug builds. Ideally, each platform should have just a single setting that is very easy to switch. When this setting is turned on, in a debug build (a) everything updates-related should be fully enabled as if in a release build (see UpdatesPackage and ExpoUpdatesReactDelegateHandler), and (b) RN should generate a bundle.
# How

For iOS, add the environment variable

```sh
export EX_UPDATES_NATIVE_DEBUG=1
```

and then reinstall pods. The podspecs now detect this environment variable, and makes changes so that the debug code that would normally load from the Metro packager is bypassed, and the app is built with the EXUpdates bundle and other dependencies needed to load updates from EAS.

# Test Plan

Manually tested against EAS.

I will investigate the possibility of building the updates E2E test app this way so that the debug build flows can be tested fully in CI.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-modules-core/ExpoModulesCore.podspec [hunk 1]
```diff
@@ -43,7 +43,9 @@ Pod::Spec.new do |s|
     s.dependency 'RCT-Folly', folly_version
   end
 
-  if !$ExpoUseSources&.include?(package['name']) && ENV['EXPO_USE_SOURCE'].to_i == 0 && File.exist?("ios/#{s.name}.xcframework") && Gem::Version.new(Pod::VERSION) >= Gem::Version.new('1.10.0')
+  ex_updates_native_debug = ENV['EX_UPDATES_NATIVE_DEBUG'] == '1'
+
+  if !ex_updates_native_debug && !$ExpoUseSources&.include?(package['name']) && ENV['EXPO_USE_SOURCE'].to_i == 0 && File.exist?("ios/#{s.name}.xcframework") && Gem::Version.new(Pod::VERSION) >= Gem::Version.new('1.10.0')
```
Comment (tsapeta) [inline packages/expo-modules-core/ExpoModulesCore.podspec:48]: Why is that needed in `expo-modules-core`? Anyway, you can revert that — we don't prebuild this package.

File: packages/expo/scripts/autolinking.rb [hunk 2]
```diff
@@ -30,4 +30,25 @@ def expo_patch_react_imports!(installer, options = {})
...
+    .map{ |t| t.user_project }
+    .uniq{ |p| p.path }
+    .push(installer.pods_project)
+
+  ex_updates_native_debug = ENV['EX_UPDATES_NATIVE_DEBUG'] == '1'
+
+  projects.each do |project|
+    project.build_configurations.each do |config|
+      if ex_updates_native_debug
+        config.build_settings['OTHER_CFLAGS'] = "$(inherited) -DEX_UPDATES_NATIVE_DEBUG=1"
```
Comment (tsapeta) [inline packages/expo/scripts/autolinking.rb:47]: This may reset flags set by other scripts, you should rather concatenate this setting with the previous value (`inherited` has different meaning in Xcode's build settings)

File: packages/expo/scripts/autolinking.rb [hunk 3]
```diff
@@ -30,4 +30,25 @@ def expo_patch_react_imports!(installer, options = {})
...
+      else
+        config.build_settings['OTHER_CFLAGS'] = "$(inherited)"
```
Comment (tsapeta) [inline packages/expo/scripts/autolinking.rb:49]: I don't quite understand why `else` and reseting existing flags are needed here

File: packages/expo/scripts/autolinking.rb [hunk 4]
```diff
@@ -30,4 +30,25 @@ def expo_patch_react_imports!(installer, options = {})
   end
 
   Expo::ReactImportPatcher.new(installer, options).run!
+  expo_patch_for_updates_debug!(installer)
+end
+
+def expo_patch_for_updates_debug!(installer)
+  projects = installer.aggregate_targets
+    .map{ |t| t.user_project }
+    .uniq{ |p| p.path }
```
Comment (tsapeta) [inline packages/expo/scripts/autolinking.rb:39]:
```suggestion
    .map { |t| t.user_project }
    .uniq { |p| p.path }
```

File: packages/expo/Expo.podspec [hunk 5]
```diff
@@ -18,4 +18,12 @@ Pod::Spec.new do |s|
   s.dependency 'ExpoModulesCore'
 
   s.source_files = 'ios/**/*.{h,m,swift}'
+
+
+  ex_updates_native_debug = ENV['EX_UPDATES_NATIVE_DEBUG'] == '1'
+  if ex_updates_native_debug
```
Comment (tsapeta) [inline packages/expo/Expo.podspec:24]:
```suggestion
  if ENV['EX_UPDATES_NATIVE_DEBUG'] == '1'
```

File: packages/expo/scripts/autolinking.rb (hunk 3 above)
Comment (douglowder) [reply packages/expo/scripts/autolinking.rb:49]: Having the else clause allows you to go back to normal debugging by removing the environment variable and doing pod install again.

Comment (tsapeta) [reply packages/expo/scripts/autolinking.rb:49]: Ah, I see. Check out https://github.com/expo/expo/blob/7aebe42601d6fadd98607fb563499c60bbfdf47a/packages/expo-modules-autolinking/scripts/ios/project_integrator.rb#L105 where we do a similar thing with Swift compiler flags. It removes the flag with `gsub` method.

File: packages/expo-modules-core/ExpoModulesCore.podspec (hunk 1 above)
Comment (douglowder) [reply packages/expo-modules-core/ExpoModulesCore.podspec:48]: If we are not prebuilding the package, maybe just remove this clause entirely and just include source files?

Comment (tsapeta) [reply packages/expo-modules-core/ExpoModulesCore.podspec:48]: Actually we didn't make the decision what to do with prebuilds yet. Kudo is still figuring out if there is an option to prebuild them even if Swift code is there.

File: packages/expo-updates/ios/EXUpdates.podspec [hunk 6]
```diff
@@ -30,17 +30,20 @@ Pod::Spec.new do |s|
     'SWIFT_COMPILATION_MODE' => 'wholemodule'
   }
 
-  if !$ExpoUseSources&.include?(package['name']) && ENV['EXPO_USE_SOURCE'].to_i == 0 && File.exist?("#{s.name}.xcframework") && Gem::Version.new(Pod::VERSION) >= Gem::Version.new('1.10.0')
+  ex_updates_native_debug = ENV['EX_UPDATES_NATIVE_DEBUG'] == '1'
```
Comment (Kudo) [inline packages/expo-updates/ios/EXUpdates.podspec:33]: i have some questions for the implementations. if this is only for expo-updates, why should we change expo, expo-modules-core, and autolinking?

- podspec supports [`compiler_flags`](https://guides.cocoapods.org/syntax/podspec.html#compiler_flags) to add defines, so we don't have to do it through autolinking.
- we could also add `pod_target_xcconfig.OTHER_SWIFT_FLAGS` if the definitions should be in swift code.

i hope we could scope the changes inside expo-updates and keeps it independent from other modules.

Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates.podspec:33]: I will attempt to scope the changes to expo-updates only, and see if I can make that work.

**Comment by douglowder** [comment]
Closing this and will rework so that only changes in the `expo-updates` module are made.
